File: src/algorithms/trees/manipulation/delete-leaves-with-value/sources/delete-leaves-with-value_test.py
import importlib
import sys
import os
from collections import deque

sys.path.insert(0, os.path.dirname(__file__))
module = importlib.import_module("delete-leaves-with-value")
BinaryNode = module.BinaryNode
delete_leaves_with_value = module.delete_leaves_with_value


def make_node(value, left=None, right=None):
    node = BinaryNode(value)
    node.left = left
    node.right = right
    return node


def collect_level_order(root):
    if root is None:
        return []
    result = []
    queue = deque([root])
    while queue:
        current = queue.popleft()
        result.append(current.value)
        if current.left:
            queue.append(current.left)
        if current.right:
            queue.append(current.right)
    return result


def test_single_node_that_is_target_returns_none():
    root = make_node(1)
    assert delete_leaves_with_value(root, 1) is None


def test_no_matching_leaf():
    root = make_node(1, make_node(2), make_node(3))
    result = delete_leaves_with_value(root, 9)
    assert collect_level_order(result) == [1, 2, 3]


def test_deletes_leaf_with_target():
    root = make_node(1, make_node(2), make_node(3))
    result = delete_leaves_with_value(root, 2)
    assert result.left is None
    assert result.right.value == 3


def test_cascades_deletion():
    root = make_node(1, make_node(2))
    result = delete_leaves_with_value(root, 2)
    assert result.value == 1
    assert result.left is None


if __name__ == "__main__":
    test_single_node_that_is_target_returns_none()
    test_no_matching_leaf()
    test_deletes_leaf_with_target()
    test_cascades_deletion()
    print("All tests passed!")

File: src/algorithms/trees/manipulation/delete-leaves-with-value/__tests__/delete-leaves-with-value_test.rs
include!("../sources/delete-leaves-with-value.rs");

#[cfg(test)]
mod tests {
    use super::*;

    fn make_node(value: i32, left: Option<Box<BinaryNode>>, right: Option<Box<BinaryNode>>) -> Option<Box<BinaryNode>> {
        Some(Box::new(BinaryNode { value, left, right }))
    }

    fn leaf(value: i32) -> Option<Box<BinaryNode>> {
        make_node(value, None, None)
    }

    #[test]
    fn test_single_target_node_returns_none() {
        let root = leaf(1);
        assert!(delete_leaves_with_value(root, 1).is_none());
    }

    #[test]
    fn test_no_matching_leaf_unchanged() {
        let root = make_node(1, leaf(2), leaf(3));
        let result = delete_leaves_with_value(root, 9);
        assert_eq!(result.as_ref().unwrap().value, 1);
        assert!(result.as_ref().unwrap().left.is_some());
        assert!(result.as_ref().unwrap().right.is_some());
    }

    #[test]
    fn test_deletes_leaf_with_target() {
        let root = make_node(1, leaf(2), leaf(3));
        let result = delete_leaves_with_value(root, 2);
        assert!(result.as_ref().unwrap().left.is_none());
        assert_eq!(result.as_ref().unwrap().right.as_ref().unwrap().value, 3);
    }

    #[test]
    fn test_cascades_deletion() {
        let root = make_node(1, leaf(2), None);
        let result = delete_leaves_with_value(root, 2);
        assert_eq!(result.as_ref().unwrap().value, 1);
        assert!(result.as_ref().unwrap().left.is_none());
    }
}
